src/cpp/session/modules/SessionRmdNotebook.R

```diff
@@ -0,0 +1,186 @@
+{
+  # create a temporary file stub to send to R Markdown
+  chunkFile <- tempfile(fileext = ".Rmd")
+  chunk <- paste(paste("```{r", options, "echo=FALSE}"),
+                 content,
+                 "```\n", 
+                 sep = "\n");
+  writeLines(chunk, con = chunkFile)
+  on.exit(unlink(chunkFile), add = TRUE)
+  
+  # render chunks directly in .GlobalEnv
+  # TODO: use .rs.getActiveFrame()? use sandbox env
+  # that has .GlobalEnv as parent?
+  envir <- .GlobalEnv
+                 
+  # render the stub to the given file
+  errorMessage <- ""
+  tryCatch({
+    capture.output(rmarkdown::render(
+      input = chunkFile, 
+      output_format = rmarkdown::html_document(
+        theme = NULL,
+        highlight = NULL,
+        template = NULL, 
+        self_contained = FALSE,
+        includes = list(
+          in_header = headerFile),
+        lib_dir = libDir),
+      output_file = outputFile,
+      encoding = "UTF-8",
+      envir = envir,
+      quiet = TRUE))
+  }, error = function(e) {
+    errorMessage <<- paste("Error:", e$message)
+  })
+
+  invisible(errorMessage)
+})
+
+.rs.addFunction("createNotebook", function(input, output = NULL, envir = .GlobalEnv)
+{
+   find_chunks <- function(contents) {
+      chunkStarts <- grep("^\\s*```{", contents, perl = TRUE)
+      chunkEnds <- grep("^\\s*```\\s*$", contents, perl = TRUE)
+      chunkRanges <- Map(list, start = chunkStarts, end = chunkEnds)
+      lapply(chunkRanges, function(range) {
+         list(start = range$start,
+              end = range$end,
+              header = contents[range$start],
+              contents = contents[(range$start + 1):(range$end - 1)])
+      })
+   }
+   
+   chunk_annotation <- function(name, index, row) {
+      sprintf("\n\n<!-- rnb-chunk-%s-%s %s -->\n\n", name, index, row)
+   }
+   
+   input <- normalizePath(input, winslash = "/", mustWork = TRUE)
+   contents <- readLines(input, warn = FALSE)
+   
+   # inject placeholders so we can track chunk locations after render
+   # ensure that the comments are surrounded by newlines, as otherwise
+   # strange render errors can occur
+   modified <- contents
+   chunks <- find_chunks(modified)
+   for (i in seq_along(chunks)) {
+      startIdx <- chunks[[i]]$start
+      modified[startIdx] <- paste(
+         chunk_annotation("start", i, startIdx),
+         modified[startIdx],
+         sep = ""
+      )
+      
+      endIdx <- chunks[[i]]$end
+      modified[endIdx] <- paste(
+         modified[endIdx],
+         chunk_annotation("end", i, endIdx),
+         sep = ""
+      )
+   }
+   
+   # generate base64 encoded version of original document
+   docEncoded <- caTools::base64encode(paste(c(contents, ""), collapse = "\n"))
+   
+   # render a GitHub-flavored version of document, for injection
+   ghOutput <- tempfile("rnb-gh-file-", fileext = ".md")
+   ghFormat <- rmarkdown::github_document(html_preview = FALSE)
+   rmarkdown::render(input = input,
```

Comment:
I think we want to make sure not to actually re-render the entire document here (as it double rendering time). Rather, we can do this as a pandoc-only markdown-to-markdown transformation (basically all we want is the conversion of pandoc markdown to github flavored markdown).